File: backend/scripts/readExcel2025.js
const XLSX = require("xlsx");
const fs = require("fs");
const path = require("path");

// ─── Helpers ──────────────────────────────────────────────────────────────────

/**
 * Parse a rank/percentage string like "31722\n(83.80)"
 * Returns { rank, percentage } or null.
 */
function parseRankPercentage(val) {
  if (!val || typeof val !== "string") return null;
  if (!val.includes("(")) return null;
  try {
    const parts = val.split("\n");
    const rank = parseInt(parts[0]);
    const pctPart = parts.find((p) => p.includes("("));
    if (!pctPart) return null;
    const percentage = parseFloat(pctPart.replace("(", "").replace(")", ""));
    if (isNaN(rank) || isNaN(percentage)) return null;
    return { rank, percentage };
  } catch {
    return null;
  }
}

/**
 * Returns true if the value looks like a college+branch header:
 * "CODE - College Name\nCODE - Branch Name\nStatus: ..."
 */
function isCollegeBranchHeader(val) {
  if (!val || typeof val !== "string") return false;
  const lines = val.split("\n");
  return (
    lines.length >= 2 &&
    lines[0].includes(" - ") &&
    lines[1].includes(" - ")
  );
}

/**
 * Get non-metadata cells from a sheet as { address: value }.
 */
function getSheetCells(sheet) {
  const cells = {};
  for (const key of Object.keys(sheet)) {
    if (!key.startsWith("!")) cells[key] = sheet[key].v;
  }
  return cells;
}

/**
 * Parse a data sheet.
 * Scans all rows to find the first "category header row" where col A is
 * empty/null and cols B+ contain category strings (e.g. "NGOPENH", "TFWS").
 * Rows after the header row are data rows with "rank\n(pct)" values.
 * Handles both row-0 and row-1 header offsets.
 * Returns array of { category, rank, percentage }.
 */
function parseDataSheet(sheet) {
  const results = [];
  const data = XLSX.utils.sheet_to_json(sheet, { header: 1 });
  if (!data || data.length < 1) return results;

  // Find the category header row: col 0 is empty, at least one col 1+ is a string
  let headerRowIdx = -1;
  let categories = [];
  for (let rowIdx = 0; rowIdx < data.length; rowIdx++) {
    const row = data[rowIdx];
    if (!row) continue;
    // col 0 must be empty
    if (row[0] != null && row[0] !== "") continue;
    // look for category strings in cols 1+
    const cats = [];
    for (let col = 1; col < row.length; col++) {
      const cat = row[col];
      if (cat && typeof cat === "string" && cat.trim().length > 0) {
        cats.push({ col, category: cat.trim() });
      }
    }
    if (cats.length > 0) {
      headerRowIdx = rowIdx;
      categories = cats;
      break;
    }
  }
  if (headerRowIdx === -1 || categories.length === 0) return results;

  // Extract rank/pct from rows after the header
  for (let rowIdx = headerRowIdx + 1; rowIdx < data.length; rowIdx++) {
    const row = data[rowIdx];
    if (!row) continue;
    for (const { col, category } of categories) {
      const cellVal = row[col];
      if (!cellVal) continue;
      const parsed = parseRankPercentage(String(cellVal));
      if (parsed) {
        results.push({ category, rank: parsed.rank, percentage: parsed.percentage });
      }
    }
  }
  return results;
}

// ─── Main Parser ──────────────────────────────────────────────────────────────

/**
 * Parse a 2025-26 workbook.
 *
 * The 2025 format spreads each college+branch across MULTIPLE sequential sheets:
 *
 *   Sheet N:   1 cell A1 = "CCODE - CName\nBCODE - BName\nStatus:..."  ← HEADER
 *   Sheet N+1: 1 cell A1 = "Home District Seats"                        ← SECTION
 *   Sheet N+2: 1 cell A1 = "Home District Non-Technical..."             ← SECTION
 *   Sheet N+3: 14+ cells  (categories in row1, ranks in rows2+)        ← DATA ✅
 *   Sheet N+4: 1 cell A1 = "Other than Home District Seats"            ← SECTION
 *   Sheet N+5: DATA ✅
 *   ... next college+branch header ...
 */
function parseWorkbook2025(filePath, round, year = 2025) {
  console.log(`  Reading ${path.basename(filePath)}...`);
  const workbook = XLSX.readFile(filePath);
  const records = [];

  // Current college+branch context (set when we hit a header sheet)
  let collegeCode = null;
  let collegeName = null;
  let branchCode = null;
  let branchName = null;

  let sheetCount = 0;

  for (const sheetName of workbook.SheetNames) {
    sheetCount++;
    if (sheetCount % 1000 === 0) {
      process.stdout.write(
        `    ... ${sheetCount}/${workbook.SheetNames.length} sheets, ${records.length} records\n`
      );
    }

    const sheet = workbook.Sheets[sheetName];
    const cells = getSheetCells(sheet);
    const cellCount = Object.keys(cells).length;

    if (cellCount === 0) continue;

    if (cellCount === 1) {
      // Single-cell sheet: either a header or a section label
      const val = String(Object.values(cells)[0] || "");

      if (isCollegeBranchHeader(val)) {
        // New college+branch group begins
        const lines = val.split("\n");
        const collegeLine = lines[0].trim();
        const branchLine = lines[1].trim();

        if (collegeLine.includes(" - ") && branchLine.includes(" - ")) {
          collegeCode = collegeLine.split(" - ")[0].trim();
          collegeName = collegeLine.split(" - ").slice(1).join(" - ").trim();
          branchCode = branchLine.split(" - ")[0].trim();
          branchName = branchLine.split(" - ").slice(1).join(" - ").trim();
        }
      }
      // Section labels ("Home District Seats", etc.) — no action needed
      continue;
    }

    // Multi-cell sheet = data table
    if (collegeCode && branchCode) {
      const entries = parseDataSheet(sheet);
      for (const { category, rank, percentage } of entries) {
        records.push({
          year,
          round,
          collegeCode,
          collegeName,
          branchCode,
          branchName,
          category,
          rank,
          percentage,
        });
      }
    }
  }

  console.log(`  Done: ${sheetCount} sheets → ${records.length} records`);
  return records;
}

// ─── File mapping ─────────────────────────────────────────────────────────────

const FILE_TO_ROUND = {
  "POST_SSC_Diploma_CAP1_Cutoff.xlsx": 1,
  "POLY25_CAP_II_CUTOFF.xlsx": 2,
  "POLY_CAPIII_CUTOFF.xlsx": 3,
  "POLY_CAPIV_CUTOFF.xlsx": 4,
};

const cwd = process.cwd();
const cliFile = process.argv[2];
let filesToProcess = [];

if (cliFile) {
  const provided = path.isAbsolute(cliFile)
    ? cliFile
    : path.join(cwd, cliFile);
  if (!fs.existsSync(provided)) {
    console.error("File not found:", provided);
    process.exit(1);
  }
  const basename = path.basename(provided);
  const round = FILE_TO_ROUND[basename];
  if (!round) {
    console.error(
      `Unknown file: ${basename}. Known: ${Object.keys(FILE_TO_ROUND).join(", ")}`
    );
    process.exit(1);
  }
  filesToProcess = [{ filePath: provided, round }];
} else {
  for (const [fileName, round] of Object.entries(FILE_TO_ROUND)) {
    const filePath = path.join(cwd, fileName);
    if (fs.existsSync(filePath)) {
      filesToProcess.push({ filePath, round });
    } else {
      console.log(`Warning: not found: ${fileName}`);
    }
  }
}

if (filesToProcess.length === 0) {
  console.error("No 2025-26 Excel files found in", cwd);
  process.exit(1);
}

for (const { filePath, round } of filesToProcess) {
  console.log(`\n=== Round ${round}: ${path.basename(filePath)} ===`);
  const records = parseWorkbook2025(filePath, round, 2025);
  console.log(`  Total: ${records.length} records for round ${round}`);

  const outName = `cutoffs2025_round${round}.json`;
  const outPath = path.join(cwd, outName);
  fs.writeFileSync(outPath, JSON.stringify(records, null, 2));
  console.log(`  Wrote ${outName}`);

  const bad = records.filter(
    (r) => !r.collegeCode || !r.branchCode || !r.category || !r.rank || !r.percentage
  );
  console.log(`  Bad records: ${bad.length}`);

  const seen = new Set();
  let dups = 0;
  for (const r of records) {
    const key = `${r.collegeCode}-${r.branchCode}-${r.category}`;
    if (seen.has(key)) dups++;
    seen.add(key);
  }
  console.log(`  Duplicates: ${dups}`);
}

File: backend/scripts/readExcel2025.py
"""
readExcel2025.py  –  Parse 2025-26 CAP cutoff Excel files into JSON.
Uses openpyxl with read_only + data_only mode for speed on large files.

Usage:
  python scripts/readExcel2025.py <filename.xlsx>
  python scripts/readExcel2025.py        # auto-discover all 4 known files
"""

import sys
import os
import re
import json
from openpyxl import load_workbook

# Map known 2025-26 filenames -> round number
FILE_TO_ROUND = {
    "POST_SSC_Diploma_CAP1_Cutoff.xlsx": 1,
    "POLY25_CAP_II_CUTOFF.xlsx": 2,
    "POLY_CAPIII_CUTOFF.xlsx": 3,
    "POLY_CAPIV_CUTOFF.xlsx": 4,
}

YEAR = 2025


def parse_rank_percentage(val):
    """
    Parse a cell value like '31722\n(83.80)' or '31722\n(83.80)\n...'
    Returns (rank, percentage) or (None, None).
    """
    if not val:
        return None, None
    s = str(val).strip()
    if "(" not in s:
        return None, None
    parts = s.split("\n")
    try:
        rank = int(parts[0].strip())
    except (ValueError, IndexError):
        return None, None
    pct_part = next((p for p in parts if "(" in p), None)
    if not pct_part:
        return None, None
    try:
        percentage = float(pct_part.replace("(", "").replace(")", "").strip())
    except ValueError:
        return None, None
    if rank != rank or percentage != percentage:  # NaN check
        return None, None
    return rank, percentage


def is_category_header_row(row_values):
    """Row where col0 is empty and at least one other col has a string."""
    if not row_values:
        return False
    if row_values[0] not in (None, "", 0):
        return False
    rest = [c for c in row_values[1:] if c and str(c).strip()]
    return len(rest) > 0


def is_section_label(row_values):
    """Row with a string in col0 and everything else empty."""
    if not row_values:
        return False
    if not isinstance(row_values[0], str):
        return False
    rest = [c for c in row_values[1:] if c not in (None, "", 0)]
    return len(rest) == 0


def parse_workbook(file_path, round_num):
    print(f"  Loading {os.path.basename(file_path)} ...", flush=True)
    wb = load_workbook(file_path, read_only=True, data_only=True)
    records = []
    sheet_count = 0

    for sheet_name in wb.sheetnames:
        sheet_count += 1
        if sheet_count % 50 == 0:
            print(f"    ... processed {sheet_count} sheets, {len(records)} records so far", flush=True)

        ws = wb[sheet_name]
        rows = list(ws.values)

        if not rows or not rows[0] or not rows[0][0]:
            continue

        header_cell = str(rows[0][0]).strip()
        header_lines = header_cell.split("\n")

        if len(header_lines) < 2:
            continue

        college_line = header_lines[0].strip()
        branch_line = header_lines[1].strip()

        if " - " not in college_line or " - " not in branch_line:
            continue

        college_parts = college_line.split(" - ", 1)
        branch_parts = branch_line.split(" - ", 1)

        college_code = college_parts[0].strip()
        college_name = college_parts[1].strip()
        branch_code = branch_parts[0].strip()
        branch_name = branch_parts[1].strip()

        if not all([college_code, college_name, branch_code, branch_name]):
            continue

        current_categories = []  # list of (col_idx, category_str)

        for row in rows[1:]:
            if not row:
                continue
            row_vals = list(row)

            if is_section_label(row_vals):
                current_categories = []
                continue

            if is_category_header_row(row_vals):
                current_categories = []
                for col_idx, val in enumerate(row_vals[1:], start=1):
                    if val and str(val).strip():
                        current_categories.append((col_idx, str(val).strip()))
                continue

            if not current_categories:
                continue

            for col_idx, category in current_categories:
                if col_idx >= len(row_vals):
                    continue
                cell_val = row_vals[col_idx]
                rank, percentage = parse_rank_percentage(cell_val)
                if rank is None:
                    continue
                records.append({
                    "year": YEAR,
                    "round": round_num,
                    "collegeCode": college_code,
                    "collegeName": college_name,
                    "branchCode": branch_code,
                    "branchName": branch_name,
                    "category": category,
                    "rank": rank,
                    "percentage": percentage,
                })

    wb.close()
    print(f"  Done: {sheet_count} sheets → {len(records)} records", flush=True)
    return records


def main():
    cwd = os.getcwd()

    cli_file = sys.argv[1] if len(sys.argv) > 1 else None
    files_to_process = []

    if cli_file:
        file_path = cli_file if os.path.isabs(cli_file) else os.path.join(cwd, cli_file)
        if not os.path.exists(file_path):
            print(f"File not found: {file_path}")
            sys.exit(1)
        basename = os.path.basename(file_path)
        round_num = FILE_TO_ROUND.get(basename)
        if not round_num:
            print(f"Unknown file: {basename}. Known files: {list(FILE_TO_ROUND.keys())}")
            sys.exit(1)
        files_to_process = [(file_path, round_num)]
    else:
        for fname, rnd in FILE_TO_ROUND.items():
            fp = os.path.join(cwd, fname)
            if os.path.exists(fp):
                files_to_process.append((fp, rnd))
            else:
                print(f"Warning: not found: {fname}")

    if not files_to_process:
        print("No 2025-26 Excel files found.")
        sys.exit(1)

    for file_path, round_num in files_to_process:
        print(f"\nProcessing round {round_num}: {os.path.basename(file_path)}", flush=True)
        records = parse_workbook(file_path, round_num)

        out_name = f"cutoffs2025_round{round_num}.json"
        out_path = os.path.join(cwd, out_name)
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(records, f, indent=2)
        print(f"  Wrote {out_name} ({len(records)} records)")

        bad = [r for r in records if not all([r["collegeCode"], r["branchCode"], r["category"], r["rank"], r["percentage"]])]
        print(f"  Bad records: {len(bad)}")

        seen = set()
        dups = 0
        for r in records:
            key = f"{r['collegeCode']}-{r['branchCode']}-{r['category']}"
            if key in seen:
                dups += 1
            seen.add(key)
        print(f"  Duplicates: {dups}")


if __name__ == "__main__":
    main()
